rasterio/_pyvsi.pyx

```diff
@@ -0,0 +1,331 @@
+# cython: language_level=3, boundscheck=False
+"""Bridge between Python file-like objects and GDAL VSI.
+
+https://gdal.org/api/cpl.html#structVSIFilesystemPluginCallbacksStruct
+"""
+
+include "gdal.pxi"
+
+import logging
+from uuid import uuid4
+
+from rasterio._base cimport open_dataset
+from rasterio._err import (
+    GDALError, CPLE_OpenFailedError, CPLE_IllegalArgError, CPLE_BaseError, CPLE_AWSObjectNotFoundError)
+
+from libc.stdio cimport printf
+from libc.string cimport memcpy
+
+from rasterio._err cimport exc_wrap_int, exc_wrap_pointer, exc_wrap_vsilfile
+
+log = logging.getLogger(__name__)
+
+gdal33_version_checked = False
+gdal33_version_met = False
+
+
+# NOTE: This has to be defined outside of gdal.pxi or other C extensions will
+# try to compile C++ only code included in this header.
+cdef extern from "cpl_vsi_virtual.h":
+    cdef cppclass VSIFileManager:
+        @staticmethod
+        void* GetHandler(const char*)
```

Comment:
Note this is the only reason this has to be compiled as C++ from my basic tests. Without this `GetHandler` function though, I'd have to re-install the plugin every time the `PyVSIFileBase` class was created. I use `GetHandler` to check if the plugin is already installed. Looking at GDAL's source it looks like it creates new objects and copies the struct every time this install is done. Not a huge deal, but not the best thing either.

If we can make the plugin installation happen at import time then maybe this wouldn't matter.